core/src/main/java/org/elasticsearch/plugins/DiscoveryPlugin.java

```diff
@@ -106,4 +110,11 @@
                                                                              NetworkService networkService) {
         return Collections.emptyMap();
     }
+
+    /**
+     * Returns a consumer that validate the initial join cluster state. The validator, unless <code>null</code> is called exactly once per
+     * join attempt but might be called multiple times during the lifetime of a node. Validators are expected to throw a
+     * {@link IllegalStateException} if the incoming cluster states is invalid.
```

Comment:
nit - since we call it both on the master and the joining node, I think we should say "if the node and the cluster state are incompatible"